File: backend/app/schemas/students.py
from pydantic import BaseModel, EmailStr, Field
from typing import Literal, Optional, Dict 
from datetime import date

class OriginalScore(BaseModel):
    original_score: int
    reading_score: int
    math_score: int

class LearningTargets(BaseModel):
    daily_goal: int
    weekly_goal: int

# student base for bronze and silver users
class StudentBase(BaseModel):
    clerk_id: str
    first_name: str 
    last_name: str
    email: EmailStr

    school: str
    state: str
    grade_level: str

    original_score: OriginalScore
    dream_score: int
    test_date: Optional[date] = None

    subscription: str
    learning_targets: LearningTargets


class RegisterStudents(StudentBase):
    referral: str


class IncomingStudentInfo(StudentBase):
    updated_at: str 


class IncomingStudentAnalytics(BaseModel):
    session_id: str
    clerk_id: str
    type: str
    topic: str 
    reading_mastery: Dict[str, int]
    reading_topics_mastery: Dict[str, Dict[str, int]]
    math_mastery: Dict[str, int]
    math_topics_mastery: Dict[str, Dict[str, int]]
    weak_subtopics: Dict[str, Dict]
    completed_at: str 

File: frontend/types/students.ts
// User Registration Information
export type UserRegistration = {
    clerk_id: string
    first_name: string
    last_name: string
    email: string

    school: string
    state: string
    grade_level: string
    original_score: {
        original_score: number
        reading_score: number
        math_score: number
    }
    dream_score: number
    test_date: string

    subscription: "free" | "starter" | "advanced"
    referral: string
    learning_targets: {
        daily_goal: number
        weekly_goal: number
    }
}